### PR TITLE
broadcom-bt-firmware: init at 12.0.1.1011

### LINKED ISSUE #21797: Broadcom based Bluetooth does not work on Dell XPS 15 9550
## Issue description
The manual solution for Ubuntu here should be adapted and automated such that it runs out of the box: http://askubuntu.com/questions/632336/bluetooth-broadcom-43142-isnt-working-on-ubuntu

```
# dmesg
[  947.674489] Bluetooth: hci0: BCM: chip id 102
[  947.690495] Bluetooth: hci0: BCM20703A1
[  947.691487] Bluetooth: hci0: BCM (001.001.005) build 0000
[  947.691533] bluetooth hci0: Direct firmware load for brcm/BCM-0a5c-6410.hcd failed with error -2
[  947.691534] Bluetooth: hci0: BCM: Patch brcm/BCM-0a5c-6410.hcd not found
```

### Steps to reproduce
See URL above

## Technical details
16.09

### PR DESCRIPTION
**Pull request by zraexy**
###### Motivation for this change

Add Broadcomm's WIDCOMM® bluetooth firmware

`broadcom-bt-firmware` can be renamed if need be. I named it that simply because it contains firmware for multiple devices and `broadcom-bt-firmware` is `bt-fw-converter`'s [repository](https://github.com/winterheart/broadcom-bt-firmware)'s name. (It also amusingly contains the firmware and the license that is being violated by redistributing the firmware.)
Adding `broadcom-bt-firmware` to `NixOS`'s `hardware.firmware` should work all with kernels except `4.4`. Making it work for `4.4` would require moving the firmware for each device into a separate package. in order for the firmware to be loaded it must have the non unique name:`BCM.hcd`.

Fixes #21797

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/os-specific/linux/firmware/bt-fw-converter/default.nix [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  unpackCmd = ''
+    mkdir -p ${name}
+    cp $src ${name}/bt-fw-converter.pl
+  '';
+
+  installPhase = ''
+    mkdir -p $out/bin
+    cp bt-fw-converter.pl $out/bin/bt-fw-converter
+    chmod +x $out/bin/bt-fw-converter
```
Comment (Mic92) [inline pkgs/os-specific/linux/firmware/bt-fw-converter/default.nix]: these three lines can be replaced by a single:

```
install -D -m755 bt-fw-converter.pl $out/bin/bt-fw-converter
```

**Comment by Mic92** [comment]
Will there any negative side-effects for 4.4, except the firmware would be not loaded?

**Comment by zraexy** [comment]
@Mic92 It should just do nothing for `4.4`. As I understand it, the naming convention that comes out of `bt-fw-converter` is used by kernels < `4.2`. `BCM.hcd` is used by `4.2`-`4.7` (The only kernel NixOS has in this range is `4.4`) and the format that the symlinks use is for kernels > `4.7`.

**Comment by Mic92** [comment]
Thanks!

**Comment by bkchr** [comment]
Hi, do I need to enable any settings for getting the broadcom firmware loaded?
I have added the following to my configuration.nix:
` hardware = {
     firmware = [pkgs.broadcom-bt-firmware];
     enableAllFirmware = true;
  };`

**Comment by Mic92** [comment]
@bkchr not that I am aware of. What is the output `nixos-version` on your system?

**Comment by Mic92** [comment]
I have not the hardware to test this, but @zraexy may could help.

**Comment by bkchr** [comment]
17.09pre107438.d5e18499d9 (Hummingbird)

The firmware is installed and I rebooted.
But I always thought that you should see loading the firmware in the kernel log, but I don't see it.

`dmesg | grep -i blue
[    2.278007] usb 1-8: Product: Bluetooth USB Host Controller
[    2.748376] Bluetooth: Core ver 2.22
[    2.748386] Bluetooth: HCI device and connection manager initialized
[    2.748388] Bluetooth: HCI socket layer initialized
[    2.748390] Bluetooth: L2CAP socket layer initialized
[    2.748395] Bluetooth: SCO socket layer initialized
[    2.792355] Bluetooth: hci0: BCM: chip id 102 build 0683
[    2.793350] Bluetooth: hci0: BCM: product 05ac:8290
[    2.809357] Bluetooth: hci0: nixos
[    2.836831] input: Broadcom Corp. Bluetooth USB Host Controller as /devices/pci0000:00/0000:00:14.0/usb1/1-8/1-8:1.0/0003:05AC:8290.0001/input/input9
[    2.888366] hid-generic 0003:05AC:8290.0001: input,hidraw1: USB HID v1.11 Keyboard [Broadcom Corp. Bluetooth USB Host Controller] on usb-0000:00:14.0-8/input0
[    2.888457] input: Broadcom Corp. Bluetooth USB Host Controller as /devices/pci0000:00/0000:00:14.0/usb1/1-8/1-8:1.1/0003:05AC:8290.0002/input/input11
[    2.941103] hid-generic 0003:05AC:8290.0002: input,hidraw3: USB HID v1.11 Mouse [Broadcom Corp. Bluetooth USB Host Controller] on usb-0000:00:14.0-8/input1
[    4.556855] Bluetooth: BNEP (Ethernet Emulation) ver 1.3
[    4.556859] Bluetooth: BNEP socket layer initialized
[   13.059911] Bluetooth: RFCOMM TTY layer initialized
[   13.059917] Bluetooth: RFCOMM socket layer initialized
[   13.059922] Bluetooth: RFCOMM ver 1.11
`

**Comment by Mic92** [comment]
Do you run the default kernel version?

**Comment by bkchr** [comment]
I would say yes? Did not build or configured a special kernel.

uname -a:
Linux nixos 4.9.28 #1-NixOS SMP Sun May 14 12:00:37 UTC 2017 x86_64 GNU/Linux

**Comment by Mic92** [comment]
Actually the logs seems fine. Does it not work for you?

**Comment by Mic92** [comment]
Could it be, that you rather require this package? https://github.com/NixOS/nixos-hardware/blob/master/apple/macbook-pro-10-1.nix#L17

**Comment by bkchr** [comment]
No it works for me, also worked before adding the files. But the connection to my bluetooth headphones is sometimes really shitty, I thought that the drivers would improve the bluetooth experience. 

I will try your other suggestion later.

**Comment by zraexy** [comment]
Your device is not one of the ones that uses this firmware. See: https://github.com/winterheart/broadcom-bt-firmware/blob/master/DEVICES.md I believe your device uses firmware from [`firmwareLinuxNonfree`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/firmware/firmware-linux-nonfree/default.nix) which has some Broadcom firmware. (enableAllFirmware also adds this)

**Pull request merged**